ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -26,297 +17,401 @@ const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
     }
+}
 
-    /// Returns true if the instruction is short.
-    fn is_short(&self) -> bool {
-        self.0 & INSTR_ISSHORT_MASK != 0
+impl LoadArgInstruction {
+    pub(crate) fn new() -> LoadArgInstruction {
+        Self {}
     }
+}
 
-    /// Returns the opcode.
-    fn opcode(&self) -> OpCode {
-        debug_assert!(self.is_short());
-        OpCode::from((self.0 & SHORT_INSTR_OPCODE_MASK) >> SHORT_INSTR_OPCODE_SHIFT)
-    }
+/// The operands for a [Instruction::Call]
+///
+/// # Semantics
+///
+/// Perform a call to an external or AOT function.
+#[derive(Debug)]
+pub struct CallInstruction {
+    /// The callee.
+    target: FuncIndex,
+    /// The first argument to the call, if present. Undefined if not present.
+    arg1: PackedOperand,
+    /// Extra arguments, if the call requires more than a single argument.
+    extra: ExtraArgsIndex,
+}
 
-    /// For short instrucitons, return the bit offset of the specified short operand.
-    fn short_operand_bit_off(&self, index: u64) -> u64 {
-        debug_assert!(self.is_short());
-        debug_assert!(index < SHORT_INSTR_MAX_OPERANDS);
-        SHORT_INSTR_FIRST_OPERAND_SHIFT + SHORT_OPERAND_SIZE * index
+impl fmt::Display for CallInstruction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "LoadArg")
     }
+}
 
-    /// Returns the specified operand.
-    fn operand(&self, index: u64) -> Operand {
-        if self.is_short() {
-            // Shift operand down the the LSB.
-            let op = self.0 >> self.short_operand_bit_off(index);
-            // Then mask it out.
-            Operand::Short(ShortOperand(op & SHORT_OPERAND_MASK))
-        } else {
-            todo!()
+impl CallInstruction {
+    pub(crate) fn new(m: &mut Module, target: FuncIndex, args: &[Operand]) -> CallInstruction {
+        let mut arg1 = PackedOperand::default();
+        let mut extra = ExtraArgsIndex::default();
+
+        if args.len() >= 1 {
+            arg1 = PackedOperand::new(&args[0]);
+        }
+        if args.len() >= 2 {
+            extra = m.push_extra_args(&args[1..]);
+        }
+        Self {
+            target,
+            arg1,
+            extra,
         }
     }
 
-    /// Create a new `Load` instruction.
-    ///
-    /// ## Operands
-    ///
-    /// - `<ptr>`:  The pointer to load from.
+    /// Fetch the operand at the specified index.
     ///
-    /// ## Semantics
-    ///
-    /// Return the value obtained by dereferencing the operand (which must be pointer-typed).
-    pub(crate) fn create_load(op: Operand) -> Self {
-        if op.is_short() {
-            let mut instr = Instruction::new_short(OpCode::Load);
-            instr.set_short_operand(op, 0);
-            instr
+    /// It is undefined behaviour to provide an out-of-bounds index.
+    pub(crate) fn operand(&self, m: &Module, idx: usize) -> Option<Operand> {
+        #[cfg(debug_assertions)]
+        {
+            let func = m.get_func(self.target);
```

Comment:
I'm not sure there's a readable one-liner here. But maybe something like:

```rust
match m.get_type(m.get_func(self.target).tyidx) {
  Type::Func(ft) => assert!(ft.num_args > idx),
  _ => panic!()
}
```

is clearer?